imgui_demo.cpp

```diff
@@ -1580,6 +1580,60 @@ static void ShowDemoWindowWidgets()
             ImGui::Unindent();
         }
 
+        {
+            ImGui::BulletText("Drag and drop to reorder items");
+            ImGui::Indent();
+
+            static bool horizontalList = false;
+            static ImVector<const char*> items;
+            if (items.empty())
+            {
+                items.push_back("Red");
+                items.push_back("Green");
+                items.push_back("Blue");
+                items.push_back("Orange");
+                items.push_back("Black");
+                items.push_back("White");
+            }
+
+            ImGui::Checkbox("Horizontal list", &horizontalList);
+
+            if (const ImGuiPayload* dropped = ImGui::AcceptReorderDropPayload("color", horizontalList ? ImGuiDragDropFlags_ReorderHorizontal : 0))
+            {
+                // A special case for moving items to front of the list.
+                const char* color = NULL;
+                memcpy(&color, dropped->Data,sizeof(color));
+                items.find_erase(color);
+                items.insert(items.begin(), color);
+            }
+
+            for (int i = 0; i < items.size(); i++)
+            {
+                const char* color = items[i];
+                ImGui::TextUnformatted(color);
+                if (horizontalList)
+                    ImGui::SameLine();
+                if (ImGui::BeginDragDropSource(ImGuiDragDropFlags_SourceAllowNullID))
+                {
+                    ImGui::SetDragDropPayload("color", &color, sizeof(color));
+                    ImGui::TextUnformatted(color);
+                    ImGui::EndDragDropSource();
+                }
+
+                if (const ImGuiPayload* dropped = ImGui::AcceptReorderDropPayload("color", horizontalList ? ImGuiDragDropFlags_ReorderHorizontal : 0))
+                {
+                    // Move item after this one.
+                    memcpy(&color, dropped->Data,sizeof(color));
+                    int prev_i = items.index_from_ptr(items.find(color));
+                    items.find_erase(color);
+                    items.insert(items.begin() + i + (prev_i > i ? 1 : 0), color);
+                }
```

Comment:
Suggestion: I'd aim to move the block code above in the loop, maybe start the loop with -1 or end with size()+1 and skip displaying the item but always display the reorder target.

Reply:
Would be nice is the demo also implemented the other form of dragging (which doesn't use the drag and drop api)